packages/core/src/domain/internalMonitoring.ts

```diff
@@ -96,27 +95,31 @@ export function resetInternalMonitoring() {
   monitoringConfiguration.batch = undefined
 }
 
-export function monitored(_: any, __: string, descriptor: PropertyDescriptor) {
-  const originalMethod = descriptor.value
-  descriptor.value = function () {
-    const decorated = (monitoringConfiguration.batch ? monitor(originalMethod) : originalMethod) as Function
-    return decorated.apply(this, arguments)
-  }
+export function monitored<T extends (...params: any[]) => unknown>(
+  _: any,
+  __: string,
+  descriptor: TypedPropertyDescriptor<T>
+) {
+  const originalMethod = descriptor.value!
+  descriptor.value = (function (this: any, ...args: Parameters<T>) {
+    const decorated = monitoringConfiguration.batch ? monitor(originalMethod) : originalMethod
+    return decorated.apply(this, args) as ReturnType<T>
+  } as any) as T
 }
 
-export function monitor<T extends Function>(fn: T): T {
-  return (function (this: any) {
+export function monitor<R>(fn: (...args: any[]) => R) {
+  return (function (this: any, ...args: any[]) {
     try {
-      return fn.apply(this, arguments)
+      return fn.apply(this, args)
     } catch (e) {
```

Comment:
This change is not free. It is transpiled like this:

```js
export function monitor(fn) {
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        try {
            return fn.apply(this, args);
        }
```

This may have a small performance impact. What about ignoring this rule in perf-critical functions like `monitor`?